compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -808,6 +748,69 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         }
     }
 
+    fn suggest_remove_semi_or_return_binding(
+        &self,
+        err: &mut Diagnostic,
+        first_id: Option<hir::HirId>,
+        first_ty: Ty<'tcx>,
+        first_span: Span,
+        second_id: Option<hir::HirId>,
+        second_ty: Ty<'tcx>,
+        second_span: Span,
+    ) {
+        let semicolon =
+            if let Some(first_id) = first_id
+                && let hir::Node::Block(blk) = self.tcx.hir().get(first_id)
+                && let Some(remove_semicolon) = self.could_remove_semicolon(blk, second_ty)
+            {
+                Some(remove_semicolon)
+            } else if let Some(second_id) = second_id
+                && let hir::Node::Block(blk) = self.tcx.hir().get(second_id)
+                && let Some(remove_semicolon) = self.could_remove_semicolon(blk, first_ty)
+            {
+                Some(remove_semicolon)
+            } else {
+                None
+            };
```

Comment:
This large if statement could use an iterative approach. Something like `[(first_id, first_ty), (second_id, second_ty)]`

Also are the arguments to `could_remove_semicolon` meant to be swapped?

Reply:
What do you mean meant to be swapped?

Reply:
The first statement checks if we can remove a semicolon from the "then" branch of the if, comparing it against the "else" expected ty. The second does the opposite.